File: VB/NonPersistentObjectsDemo.Module/BusinessObjects/AuxiliaryObjects.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.DC
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.Validation

Namespace NonPersistentObjectsDemo.Module.BusinessObjects

	<DefaultClassOptions>
	<DefaultListViewOptions(True, NewItemRowPosition.Top)>
	<DefaultProperty("Subject")>
	<DevExpress.ExpressApp.DC.DomainComponent>
	Public Class Message
		Inherits NonPersistentObjectBase

'INSTANT VB NOTE: The field id was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private id_Conflict As Integer
		<Browsable(False)>
		<DevExpress.ExpressApp.Data.Key>
		Public ReadOnly Property ID() As Integer
			Get
				Return id_Conflict
			End Get
		End Property
'INSTANT VB NOTE: The variable id was renamed since Visual Basic does not handle local variables named the same as class members well:
		Public Sub SetKey(ByVal id_Conflict As Integer)
			Me.id_Conflict = id_Conflict
		End Sub
		Private _Sender As Account
		Public Property Sender() As Account
			Get
				Return _Sender
			End Get
			Set(ByVal value As Account)
				SetPropertyValue(NameOf(Sender), _Sender, value)
			End Set
		End Property
		Private _Recepient As Account
		Public Property Recepient() As Account
			Get
				Return _Recepient
			End Get
			Set(ByVal value As Account)
				SetPropertyValue(NameOf(Recepient), _Recepient, value)
			End Set
		End Property
		Private _Subject As String
		Public Property Subject() As String
			Get
				Return _Subject
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(Subject), _Subject, value)
			End Set
		End Property
		Private _Body As String
		<FieldSize(-1)>
		Public Property Body() As String
			Get
				Return _Body
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(Body), _Body, value)
			End Set
		End Property
	End Class

	<DefaultClassOptions>
	<DefaultListViewOptions(True, NewItemRowPosition.Top)>
	<DefaultProperty("PublicName")>
	<DevExpress.ExpressApp.DC.DomainComponent>
	Public Class Account
		Inherits NonPersistentObjectBase

'INSTANT VB NOTE: The field userName was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private userName_Conflict As String
		'[Browsable(false)]
		<DevExpress.ExpressApp.ConditionalAppearance.Appearance("", Enabled := False, Criteria := "Not IsNewObject(This)")>
		<RuleRequiredField>
		<DevExpress.ExpressApp.Data.Key>
		Public Property UserName() As String
			Get
				Return userName_Conflict
			End Get
			Set(ByVal value As String)
				userName_Conflict = value
			End Set
		End Property
'INSTANT VB NOTE: The variable userName was renamed since Visual Basic does not handle local variables named the same as class members well:
		Public Sub SetKey(ByVal userName_Conflict As String)
			Me.userName_Conflict = userName_Conflict
		End Sub
'INSTANT VB NOTE: The field publicName was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private publicName_Conflict As String
		Public Property PublicName() As String
			Get
				Return publicName_Conflict
			End Get
			Set(ByVal value As String)
				SetPropertyValue(NameOf(PublicName), publicName_Conflict, value)
			End Set
		End Property
	End Class


	Public Class PostOfficeFactory
		Inherits NonPersistentObjectFactoryBase

		Private ReadOnly Property Storage() As PostOfficeClient
			Get
				Return GlobalServiceProvider(Of PostOfficeClient).GetService()
			End Get
		End Property
		Private isLoading As Boolean = False
		Private objectMap As ObjectMap
		Public Sub New(ByVal objectMap As ObjectMap)
			Me.objectMap = objectMap
		End Sub
		Public Overrides Function GetObjectByKey(ByVal objectType As Type, ByVal key As Object) As Object
			If key Is Nothing Then
				Throw New ArgumentNullException(NameOf(key))
			End If
			Dim mapping As DataStoreMapping
			If Storage.Mappings.TryGetValue(objectType, mapping) Then
				Return WrapLoading(Function()
					Dim loader = New DataStoreObjectLoader(Storage.Mappings, Storage.DataStore, objectMap)
					Return loader.LoadObjectByKey(objectType, key)
				End Function)
			End If
			Throw New NotImplementedException()
		End Function
		Public Overrides Function GetObjects(ByVal objectType As Type, ByVal criteria As CriteriaOperator, ByVal sorting As IList(Of DevExpress.Xpo.SortProperty)) As IEnumerable
			Dim mapping As DataStoreMapping
			If Storage.Mappings.TryGetValue(objectType, mapping) Then
				Return WrapLoading(Function()
					Dim loader = New DataStoreObjectLoader(Storage.Mappings, Storage.DataStore, objectMap)
					Return loader.LoadObjects(objectType, criteria)
				End Function)
			End If
			Throw New NotImplementedException()
		End Function
		Private Function WrapLoading(Of T)(ByVal doer As Func(Of T)) As T
			If isLoading Then
				Throw New InvalidOperationException()
			End If
			isLoading = True
			Try
				Return doer.Invoke()
			Finally
				isLoading = False
			End Try
		End Function
		Public Overrides Sub SaveObjects(ByVal toInsert As ICollection, ByVal toUpdate As ICollection, ByVal toDelete As ICollection)
			Dim saver = New DataStoreObjectSaver(Storage.Mappings, Storage.DataStore)
			saver.SaveObjects(toInsert, toUpdate, toDelete)
		End Sub
	End Class
End Namespace

File: CS/NonPersistentObjectsDemo.Module/BusinessObjects/AuxiliaryObjects.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.DC;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.Validation;
using DevExpress.Xpo;
namespace NonPersistentObjectsDemo.Module.BusinessObjects {

    [DefaultClassOptions]
    [DefaultListViewOptions(true, NewItemRowPosition.Top)]
    [DefaultProperty("Subject")]
    [DevExpress.ExpressApp.DC.DomainComponent]
    public class Message : NonPersistentObjectBase {
        private int id;
        [Browsable(false)]
        [DevExpress.ExpressApp.Data.Key]
        public int ID {
            get { return id; }
        }
        public void SetKey(int id) {
            this.id = id;
        }
        private Account _Sender;
        public string SenderPublicName { get => _Sender.PublicName;
            set { throw new Exception("Set not implemented"); }
        }
        public Account Sender {
            get { return _Sender; }
            set { SetPropertyValue(nameof(Sender), ref _Sender, value); }
        }
        private Account _Recepient;
        public Account Recepient {
            get { return _Recepient; }
            set { SetPropertyValue(nameof(Recepient), ref _Recepient, value); }
        }
        private string _Subject;
        public string Subject {
            get { return _Subject; }
            set { SetPropertyValue<string>(nameof(Subject), ref _Subject, value); }
        }
        private string _Body;
        [FieldSize(-1)]
        public string Body {
            get { return _Body; }
            set { SetPropertyValue<string>(nameof(Body), ref _Body, value); }
        }
    }

    [DefaultClassOptions]
    [DefaultListViewOptions(true, NewItemRowPosition.Top)]
    [DefaultProperty("PublicName")]
    [DevExpress.ExpressApp.DC.DomainComponent]
    public class Account : NonPersistentObjectBase {
        private string userName;
        //[Browsable(false)]
        [DevExpress.ExpressApp.ConditionalAppearance.Appearance("", Enabled = false, Criteria = "Not IsNewObject(This)")]
        [RuleRequiredField]
        [DevExpress.ExpressApp.Data.Key]
        public string UserName {
            get { return userName; }
            set { userName = value; }
        }
        public void SetKey(string userName) {
            this.userName = userName;
        }
        private string publicName;

        private BindingList<Message> _messages;
         
       // [DevExpress.ExpressApp.DC.Aggregated]
        public BindingList<Message> Messages {
            get {
                if(_messages == null) {
                    _messages = new BindingList<Message>( );  // errors here
                }
                CriteriaOperator criteria = new BinaryOperator(
                    new OperandProperty("SenderPublicName"), new OperandValue(PublicName),
                    BinaryOperatorType.Equal
                );
                //CriteriaOperator criteria = new BinaryOperator(
                //    new OperandProperty("Subject"), new OperandValue(" "),
                //    BinaryOperatorType.GreaterOrEqual
                //);

                var mgs = ObjectSpace.GetObjects<Message>(criteria);
                foreach(var m in mgs ) {
                    _messages.Add(m);
                }

                return _messages;
            }
        }

        public string PublicName {
            get { return publicName; }
            set { SetPropertyValue(nameof(PublicName), ref publicName, value); }
        }
    }


    public class PostOfficeFactory : NonPersistentObjectFactoryBase {
        private PostOfficeClient Storage => GlobalServiceProvider<PostOfficeClient>.GetService();
        private bool isLoading = false;
        private ObjectMap objectMap;
        public PostOfficeFactory(ObjectMap objectMap) {
            this.objectMap = objectMap;
        }
        public override object GetObjectByKey(Type objectType, object key) {
            if(key == null) {
                throw new ArgumentNullException(nameof(key));
            }
            if(Storage.Mappings.TryGetValue(objectType, out var mapping)) {
                return WrapLoading(() => {
                    var loader = new DataStoreObjectLoader(Storage.Mappings, Storage.DataStore, objectMap);
                    return loader.LoadObjectByKey(objectType, key);
                });
            }
            throw new NotImplementedException();
        }
        public override IEnumerable GetObjects(Type objectType, CriteriaOperator criteria, IList<DevExpress.Xpo.SortProperty> sorting) {
            if(Storage.Mappings.TryGetValue(objectType, out var mapping)) {
                return WrapLoading(() => {
                    var loader = new DataStoreObjectLoader(Storage.Mappings, Storage.DataStore, objectMap);
                    return loader.LoadObjects(objectType, criteria);
                });
            }
            throw new NotImplementedException();
        }
        private T WrapLoading<T>(Func<T> doer) {
            if(isLoading) {
                throw new InvalidOperationException();
            }
            isLoading = true;
            try {
                return doer.Invoke();
            }
            finally {
                isLoading = false;
            }
        }
        public override void SaveObjects(ICollection toInsert, ICollection toUpdate, ICollection toDelete) {
            var saver = new DataStoreObjectSaver(Storage.Mappings, Storage.DataStore);
            saver.SaveObjects(toInsert, toUpdate, toDelete);
        }
    }
}
